### ISSUE TITLE
keepalive `include`/`exclude` props don't work with `<NuxtPage>`

### ISSUE THREAD

**Issue by Ilmion** [nuxt/nuxt#15214]
### Environment

------------------------------
- Operating System: `Windows_NT`
- Node Version:     `v16.16.0`
- Nuxt Version:     `3.0.0-rc.11`
- Nitro Version:    `0.5.4`
- Package Manager:  `npm@8.11.0`
- Builder:          `vite`
- User Config:      `-`
- Runtime Modules:  `-`
- Build Modules:    `-`
------------------------------

### Reproduction

https://stackblitz.com/edit/github-xnrjjn?file=pages%2Findex.vue,pages%2Fnotkeptalive.vue

### Describe the bug

App has 2 pages, one page is set to be kept alive with 
`definePageMeta({ keepalive: true });`
the other one is not.

Changing some value in the first page, is not kept if you navigate away from it and if you come back.
Also, there is a weird display bug, as if a new mounted component of the page was added under the first one in the page before the first one being added. It make the page flicker.

If you put the same page meta in the second one (by removing the comment is the reproduction link), now it works, but the two pages are now kept alive and this is not what I want.

In base vue3, I could do something like that :
```
<router-view v-slot="{ Component }">
      <keep-alive include="index">
        <component :is="Component" />
      </keep-alive>
    </router-view>
```

**Comment by vcasy** [comment]
I hope this gets fixed sometime. 
Just checked with newest release `v3.1.1` - bug still remains (same reason described by Ilmion).

**Workaround: Use `<RouterView />` instead `<NuxtPage />`**

If you use this solution you also have to manully import `useRoute` everytime:
`import { useRoute } from 'vue-router';`

Because the auto-imported `useRoute` of nuxt will be wrong

**Comment by toniengelhardt** [comment]
Did anybody find a workaround? I see there is a hack for keepalive also implemented in Elk: 

https://github.com/elk-zone/elk/blob/4422a57f491fa2ee53ca303b9e693127caec83a0/composables/vue.ts

but for me the `onActivated` and `onDeactivated` hooks are also not triggered...

**Comment by AndrewBogdanovTSS** [comment]
@danielroe any updates on this one? Thanks

**Comment by ajh99990** [comment]
In the latest version v3.3.1, this bug still exists.

**Comment by kikuchan** [comment]
After some investigation, I've noticed that the following rules:

- `KeepAlive` only looks at the `name` of the component for include/exclude
- A page component wrapped by `NuxtPage`, is also wrapped by `RouteProvider` named `RouteProvider`

That's why the include/exclude keepalive options don't work for `NuxtPage`.

There are some solutions worth to consider I've found so far;
1. Make `KeepAlive` looks at `key` in addition to `name`, in Vue3
2. Give each `RouteProvider` component a distinct name.
3. Remove `RouteProvider` and provide `_route` reactive somehow.

I've also came up with a PoC code for 2nd solution, but bit hacky;
```ts
--- packages/nuxt/src/pages/runtime/page.ts
+++ packages/nuxt/src/pages/runtime/page.ts
@@ -36,6 +36,8 @@ export default defineComponent({
   },
   setup (props, { attrs }) {
     const nuxtApp = useNuxtApp()
+    const RouteProvider = defineComponent({ ...RouteProviderTemplate })
+
     return () => {
       return h(RouterView, { name: props.name, route: props.route, ...attrs }, {
         default: (routeProps: RouterViewSlotProps) => {
@@ -44,6 +46,8 @@ export default defineComponent({
           const key = generateRouteKey(routeProps, props.pageKey)
           const done = nuxtApp.deferHydration()

+          RouteProvider.name = String(key)
+
           const hasTransition = !!(props.transition ?? routeProps.route.meta.pageTransition ?? defaultPageTransition)
           const transitionProps = hasTransition && _mergeTransitionProps([
             props.transition,
@@ -77,12 +81,11 @@ function _mergeTransitionProps (routeProps: TransitionProps[]): TransitionProps
   return defu(..._props)
 }

-const RouteProvider = defineComponent({
-  name: 'RouteProvider',
+const RouteProviderTemplate = ({
   // TODO: Type props
   // eslint-disable-next-line vue/require-prop-types
   props: ['routeProps', 'pageKey', 'hasTransition'],
-  setup (props) {
+  setup (props: any) {
     // Prevent reactivity when the page will be rerendered in a different suspense fork
     // eslint-disable-next-line vue/no-setup-props-destructure
     const previousKey = props.pageKey
```

**Comment by danielroe** [comment]
Yes, that's a very good summary. We do need the routeProvider to ensure that the route is 'frozen' to each suspense fork until suspense resolves. Creating additional copies of RouteProvider with the name matching the route component is probably the best solution to explore (likely with WeakMap to avoid memory leak on client-side).

**Comment by BunnyMan1** [comment]
I have the same issue with <slot /> inside a layout.

**Comment by wf-soft** [comment]
I have the same problem at Nuxt3. Is there an alternative solution?

**Comment by wf-soft** [comment]
There doesn't seem to be a clear and effective solution to this problem here, and it doesn't work. Even if the basic keepalive sets up a separate page, it doesn't have any effect. I've been searching all day without any useful information about this problem.

**Comment by wf-soft** [comment]
![image](https://user-images.githubusercontent.com/62228539/232997989-2c270289-4871-433a-99ee-137e613e2b5e.png)
![image](https://user-images.githubusercontent.com/62228539/232998101-8b62b71c-9eef-4dac-8383-83f115bf456b.png)

Even if key is specified, the page cannot be cached

**Comment by wf-soft** [comment]
It seems that in order to achieve my goal, I have to use middleware to achieve the desired results.

```
export default defineNuxtRouteMiddleware((to, from) => {
  const keepArray = [
    [ "/","/:column(3d|su|cz|tt|dg|ps|cad)/:id(\\d+).html" ],
    [ "/anli", "/anli/:id.html"],
    [ "/lg", "/lg/:id.html" ]
  ]
  const fromPath = from.fullPath;
  const toPath = to.fullPath;
  const fromMatchedPath = from.matched.length > 0 ? from.matched[0].path : "";
  const toMatchedPath = to.matched.length > 0 ? to.matched[0].path : "";
  const keep = keepArray.find(x=> x.includes(toPath) || x.includes(toMatchedPath));
  if(keep)
  {
    to.meta.keepalive = true;
    if(!(fromPath === toPath ||
        (keep.includes(fromPath) && keep.includes(toMatchedPath)) ||
        (keep.includes(toPath) && keep.includes(fromMatchedPath))
    )){
      from.meta.keepalive = false;
    }
  }
  else {
    from.meta.keepalive = false;
    to.meta.keepalive = false;
  }
}
```

**Comment by soylomass** [comment]
Does anyone know how to put the KeepAlive outside the layout?
As NuxtPage, according to the docs, is put inside NuxtLayout, and the keepalive property goes in NuxtPage, it means that the keepalive is cleared each time the layout changes, because the keepalive is inside, not outside, the layout.

Is there any workaround to preserve keepalive while changing layouts?

https://prnt.sc/woqGpakd7zrS

**Comment by gyhyfj** [comment]
Is there any alternative solution for page level keep-alive?

**Comment by tenfei4** [comment]
You can use a routing middleware to do this, such as using useState to record a list, and each time you compare whether it is in the list, you can set the page you want to go to as keepalive. This example is not perfect, however, because of a bug built into nuxt3 that causes the keepalives to change for all pages every time you set any to.meta.keepalive to true.
```
export default defineNuxtRouteMiddleware((to,from)=>{
    if(useAliveList().value.indexOf(to.path)>=0){
        to.meta.keepalive=true;
    }else{
        to.meta.keepalive=false;
    }
    useAliveList('add',to.path);
});
```

**Comment by gyhyfj** [comment]
> You can use a routing middleware to do this, such as using useState to record a list, and each time you compare whether it is in the list, you can set the page you want to go to as keepalive. This example is not perfect, however, because of a bug built into nuxt3 that causes the keepalives to change for all pages every time you set any to.meta.keepalive to true.
> 
> ```
> export default defineNuxtRouteMiddleware((to,from)=>{
>     if(useAliveList().value.indexOf(to.path)>=0){
>         to.meta.keepalive=true;
>     }else{
>         to.meta.keepalive=false;
>     }
>     useAliveList('add',to.path);
> });
> ```

Thank you, what’s the meaning of  ` useAliveList('add',to.path)` ?

**Comment by tenfei4** [comment]
@gyhyfj useAliveList() is a useState of compoables that only returns a list. This instance only controls the pages in the list to be keepalive, but once you visit other pages, all the list pages will be invalid, and you need to reload the list after the second visit to achieve the function, because each time you set to.meta.keepalive will cause all pages to become true.
you can try this example.(first only vist /a,/b,/c)
```
export default defineNuxtRouteMiddleware((to,from)=>{
  if(['/a','/b','/c'].indexOf(to.path)>=0){
         to.meta.keepalive=true;
  }else{
      to.meta.keepalive=false;
  }
});
```

**Comment by tenfei4** [comment]
@gyhyfj useAliveList() is a useState of compoables that only returns a list. This instance only controls the pages in the list to be keepalive, but once you visit other pages, all the list pages will be invalid, and you need to reload the list after the second visit to achieve the function, because each time you set to.meta.keepalive will cause all pages to become true.
you can try this example.(just only visit /a,/b,/c pages and then visit a other page)
```
export default defineNuxtRouteMiddleware((to,from)=>{
  if(['/a','/b','/c'].indexOf(to.path)>=0){
         to.meta.keepalive=true;
  }else{
      to.meta.keepalive=false;
  }
});
```

**Comment by gyhyfj** [comment]
> @gyhyfj useAliveList() is a useState of compoables that only returns a list. This instance only controls the pages in the list to be keepalive, but once you visit other pages, all the list pages will be invalid, and you need to reload the list after the second visit to achieve the function, because each time you set to.meta.keepalive will cause all pages to become true. you can try this example.(just only visit /a,/b,/c pages and then visit a other page)
> 
> ```
> export default defineNuxtRouteMiddleware((to,from)=>{
>   if(['/a','/b','/c'].indexOf(to.path)>=0){
>          to.meta.keepalive=true;
>   }else{
>       to.meta.keepalive=false;
>   }
> });
> ```

Thank you. However it doesn't help in my project because I must keep one special page alive , on whitch a js sandbox and a shadow root couldn't be destroied after leaving that page. Hope this issue could be resolve by nuxt 🙏

**Comment by Megasu** [comment]
## The documentation about KeepAlive is too simple.

It can be quite challenging for developers who are unfamiliar with Vue.js, as the documentation may not be particularly user-friendly.

## Bugs
keepalive `include`/`exclude` props don't work with `<NuxtPage>`

**Comment by FuLei-hub** [comment]
same too, it is very big bug, can not control page keepalive, only all use, or all no

**Comment by yveeven** [comment]
I'd like to ask when the update will be available.

**Comment by daomiaru** [comment]
I would like to know the latest progress on this issue? It seems that the latest version of nuxt still hasn't been reso

**Comment by daomiaru** [comment]
I would like to know the latest progress on this issue? It seems that the latest version of nuxt still hasn't been resolved

**Comment by transtone** [comment]
![image](https://github.com/nuxt/nuxt/assets/46688/bf58a990-c053-47fe-a994-a81a4c666644)
![image](https://github.com/nuxt/nuxt/assets/46688/62c01486-a92b-4877-b12c-871232629c42)
`Oct 21, 2022` -> `Nov 28, 2022`  -> `Jul 17， 2023`

is there any hope to fix this `[p4-important](https://github.com/nuxt/nuxt/labels/%E2%9D%97%20p4-important)` bug?

**Comment by freezyh** [comment]
> Yes, that's a very good summary. We do need the routeProvider to ensure that the route is 'frozen' to each suspense fork until suspense resolves. Creating additional copies of RouteProvider with the name matching the route component is probably the best solution to explore (likely with WeakMap to avoid memory leak on client-side).

same too, it is very big bug, can not control page keepalive, only all use, or all no

**Comment by freezyh** [comment]
> In the latest version v3.3.1, this bug still exists.

it's a big bug!

**Comment by freezyh** [comment]
> In the latest version v3.3.1, this bug still exists.

 it is very big bug, can not control page keepalive, only all use, or all no

**Comment by freezyh** [comment]
> Yes, this is definitely on the roadmap and needs to be fixed.

 it is very big bug, can not control page keepalive, only all use, or all no

**Comment by danielroe** [comment]
@freezyh Please stop commenting. I understand you're frustrated, but please be considerate of others. Everyone subscribed to this issue will receive a notification, and you are not helping solve the issue.

This is definitely a bug we want to fix, as you can see by the fact that it is labeled `p4-important`.

**Comment by IamTyb** [comment]
我也遇到了这个问题，大家可以先用definePageMeta({
  keepalive: true,
})的方式去解决，不过要去需要去需要缓存的页面添加，在to和form页面上都添加才能用

**Comment by kongyijilafumi** [comment]
I create a example,what should i do  how to use keepalive  in  the `a` page. [example](https://stackblitz.com/edit/nuxt-starter-sxlfea?file=app.vue)

**Comment by geauser** [comment]
I know the issue is being worked on by the team of nuxt, but in the mean time, does anyone have a comprehensible guide for a workaround? I don't tried this but it does not work: https://github.com/nuxt/nuxt/issues/15214#issuecomment-1411650328

**Comment by mkarras** [comment]
> I know the issue is being worked on by the team of nuxt, but in the mean time, does anyone have a comprehensible guide for a workaround? I don't tried this but it does not work: https://github.com/nuxt/nuxt/issues/15214#issuecomment-1411650328

I think you need to disable keep-alive completely to for example not run into security issues from cached data that should not be cached. Enable keep-alive back with the excluded routes when this issue is fixed.

When using the router directly, there are issues with the Nuxt ecosystem.

**Comment by henriquevschroeder** [comment]
Hello, I'd like to ask, are there any updates from the team? Thanks in advance.

**Comment by 11003** [comment]
+1

**Comment by gjssss** [comment]
I tried it out and found that the `route-provider` component is indeed preventing the `keep-alive` from matching components by name, which is causing the functionality to fail. I've come up with two possible solutions:

1. Dynamically change the name of the `route-provider` component. I think it might be reasonable to change the component name to something based on the included page name, such as `IndexProvider`. However, changing the component name at runtime may not be ideal, but given the current issue, it could serve as a temporary solution.

2. Add support for Vue 3 `keep-alive` to match components based on a key. This solution may require intervention from the Vue team. However, it seems that this issue originated from the design of Nuxt, and waiting for Vue to address it might take some time, which could be a loss for Nuxt. I've attempted to use some hacky methods to add components to the `Cache Map` of the `keep-alive` component but haven't succeeded.

3. Is it possible to inject the provider into the user's page in some way, automatically wrapping the user's code, and ensuring that `keep-alive` uses the page component for matching (I haven't tried this, and I'm not sure if it's feasible)?

In summary, I hope that Nuxt can provide some temporary solutions for this issue (assuming that the API won't change) and include warnings or notifications about this problem in the documentation to help those who are facing similar problems.🌹

**Comment by gjssss** [comment]
@AndrewBogdanovTSS Indeed, I believe that at least for now, some temporary solutions can be used to address this issue, or the problem can be highlighted in the Nuxt documentation.

**Comment by AndrewBogdanovTSS** [comment]
@gjssss absolutely agree

**Comment by karlsbeard** [comment]
> I tried it out and found that the `route-provider` component is indeed preventing the `keep-alive` from matching components by name, which is causing the functionality to fail. I've come up with two possible solutions:
> 
> 1. Dynamically change the name of the `route-provider` component. I think it might be reasonable to change the component name to something based on the included page name, such as `IndexProvider`. However, changing the component name at runtime may not be ideal, but given the current issue, it could serve as a temporary solution.
> 2. Add support for Vue 3 `keep-alive` to match components based on a key. This solution may require intervention from the Vue team. However, it seems that this issue originated from the design of Nuxt, and waiting for Vue to address it might take some time, which could be a loss for Nuxt. I've attempted to use some Hacking methods to add components to the `Cache Map` of the `keep-alive` component but haven't succeeded.
> 3. Is it possible to inject the provider into the user's page in some way, automatically wrapping the user's code, and ensuring that `keep-alive` uses the page component for matching (I haven't tried this, and I'm not sure if it's feasible)?
> 
> In summary, I hope that Nuxt can provide some temporary solutions for this issue (assuming that the API won't change) and include warnings or notifications about this problem in the documentation to help those who are facing similar problems.🌹

nice shot

**Comment by sanba-anass** [comment]
setting app : {keepalive:true} in the nuxt config file  and disabling each page i don't want to keep alive with definePageMeta({ keepalive: false }) works for me ,is there any downside using this method thanks in advance!

**Comment by citrusjunoss** [comment]
There is an urgent need for someone to provide a temporary solution so that the problem can be addressed as soon as possible

**Comment by huoyou** [comment]
您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。

**Comment by mkarras** [comment]
> setting app : {keepalive:true} in the nuxt config file  and disabling each page i don't want to keep alive with definePageMeta({ keepalive: false }) works for me ,is there any downside using this method thanks in advance!

I think someone here already tried. The first time when a non keep-alive page is visited,the whole cache is cleared. Making the feature more or less useless.

**Comment by stenet** [comment]
@danielroe @bencodezen could you please add a reference to the related vue issue here? I tried to find it but I'm not sure which one it is.

Thanks!

**Comment by Maxbsy** [comment]
Any idea when this will be fixed? Or can you tell us about the fix plan for this problem? Is there any hope of solving this problem in the short term? If not, can you provide a temporary solution?

**Comment by kr2748** [comment]
keepalive include/exclude props don't work with
When this will be resolved?
We want to use include/exclude props.
I know you are busy and having a hard time, but I hope you can resolve this as soon as possible.

**Comment by Clarkkkk** [comment]
@Maxbsy @kr2748 It has been fixed and should be available in 3.8.2 soon.

**Comment by BoBo-Git** [comment]
> @Maxbsy @kr2748 It has been fixed and should be available in 3.8.2 soon.

I can't wait to try out this version, but its performance seems to be different from what I expected. The lifecycle of the component can be triggered normally, but it also triggers data requests on the page, making me feel like the page is being refreshed. However, if I only use `keepalive` props like  `<NuxtPage :keepalive={}>`, the data requests on the page will not be triggered. I want to know why this is happening.

**Comment by BoBo-Git** [comment]
I have made a series of attempts again, and currently this solution can achieve the goal I want.

1. Configure `keepalive` as true in `nuxt.config.ts`.
2. Remove the `keepalive` props on the `<NuxtPage>`.
3. Control `keepalive` only through the `keepalive` property in `definePageMeta`.

**Comment by bigiCrab** [comment]
I try it on NUXT 3.8.2, for page keepalive and it work.

1. set the name for the page you want to keepalive, not the name in `definePageMeta` but in the options api.
2. set `<NuxtPage :keepalive="{include: 'KeepAliveComponentName'}" />` in app.vue.

note: you can also control the key by setting the key in `definePageMeta`

**Comment by freezyh** [comment]
> I try it on NUXT 3.8.2, for page keepalive and it work.
> 
> 1. set the name for the page you want to keepalive, not the name in `definePageMeta` but in the options api.
> 2. set `<NuxtPage :keepalive="{include: 'KeepAliveComponentName'}" />` in app.vue.
> 
> note: you can also control the key by setting the
I don't know how to set the name for the page
my code:
  <NuxtPage :keepalive="{ include: 'alive1' }" />
and the page src:  pages\alive1\index.vue

It can't not work

**Comment by freezyh** [comment]
> I try it on NUXT 3.8.2, for page keepalive and it work.
> 
> 1. set the name for the page you want to keepalive, not the name in `definePageMeta` but in the options api.
> 2. set `<NuxtPage :keepalive="{include: 'KeepAliveComponentName'}" />` in app.vue.
> 
> note: you can also control the key by setting the
I don't know how to set the name for the page
my code:
  <NuxtPage :keepalive="{ include: 'alive1' }" />
and the page src:  pages\alive1\index.vue

It can't not work

**Comment by freezyh** [comment]
> I try it on NUXT 3.8.2, for page keepalive and it work.
> 
> 1. set the name for the page you want to keepalive, not the name in `definePageMeta` but in the options api.
> 2. set `<NuxtPage :keepalive="{include: 'KeepAliveComponentName'}" />` in app.vue.
> 
> note: you can also control the key by setting the key in `definePageMeta`

I don't know how to set the name for the page

my code: 
<NuxtPage :keepalive="{ include: 'alive1' }" />

the page src:

G:\study\testnuxt3\pages\alive1\index.vue

It can't work , can you privide a sameple code, thank you very much!

**Comment by bigiCrab** [comment]
@freezyh use this to add name
```
<script>
export default {
  name: 'alive1',
};
</script>
```
I also update the demo in the above comment.
hope this help!

**Comment by freezyh** [comment]
> @freezyh use this to add name
> 
> ```
> <script>
> export default {
>   name: 'alive1',
> };
> </script>
> ```
> 
> I also update the demo in the above comment. hope this help!

Thanks,but I use typescript,how to wright in typescript
`<script lang="ts" setup></script>`

**Comment by freezyh** [comment]
> @freezyh use this to add name
> 
> ```
> <script>
> export default {
>   name: 'alive1',
> };
> </script>
> ```
> 
> I also update the demo in the above comment. hope this help!

thank you very much, I had wright it by typescript
`defineOptions({ name: 'Alive1' })`
It works!

**Comment by transtone** [comment]
https://github.com/nuxt/nuxt/pull/24024#issuecomment-1870467296

**Comment by PNP-cpp** [comment]
Dynamically changing the include attribute will invalidate the cache. Who knows what happened
`

 <NuxtPage :page-key="(route) => route.meta.id" :keepalive="{ include: cacheIncludes, max: 10 }" />

`

**Comment by PNP-cpp** [comment]
Dynamically changing the include attribute will invalidate the cache. Who knows what happened
`html
<NuxtPage :page-key="(route) => route.meta.id" :keepalive="{ include: cacheIncludes, max: 10 }" />
`

**Comment by ronannnn** [comment]
> Dynamically changing the include attribute will invalidate the cache. Who knows what happened `html <NuxtPage :page-key="(route) => route.meta.id" :keepalive="{ include: cacheIncludes, max: 10 }" /> `

I have the same problem

**Comment by tenfei4** [comment]
After testing, in v3.11.1 version, has been able to solve this problem very well, here is an idea, The reason for the problem is **keepalive include** default read **from.matched[0].components.default.__name** The value is used as the name of the components, but nuxt defaults to file as the route name.
 So we can add the following code to middleware:
```
 from.matched[0].components.default.__ name = from.name;
useAliveList('add', from.name?. toString());
```
 You can replace the default component name with from.name and add from.name to a list of composables or pinia states.
Then add the following code to app.vue: 
```
<template>
	<NuxtLayout>
		<NuxtLoadingIndicator color="#3366ff" class="shadow-md shadow-blue-400" />
		<NuxtPage :keepalive="{include:alive}" :pageKey="route=>route.fullPath"/>
	</NuxtLayout>
</template>

<script setup lang="ts">
const route=useRoute(),aliveList=useAliveList();
const alive=computed(()=>aliveList.value.join(','));
//watch(()=>alive.value,()=>{ console.log(alive.value); });
</script>
```

**Comment by tenfei4** [comment]
Of course, I hope that the official can match from.matched[0].components.default.__name The default value directly to the value of form.name. That would be even more perfect.

**Comment by alih70442** [comment]
complete usage in nuxt v3.17.5:
```
<NuxtLayout>
  <NuxtPage :keepalive="{ include: 'store.category', max: 3 }" />
</NuxtLayout>
```
also I had to add this code to my page component:
```
<script setup>
defineOptions({
  name: 'store.category',
});
</script>
```
